02-library/cs506/sim.py

```
@@ -1,13 +1,44 @@
 def euclidean_dist(x, y):
-    raise NotImplementedError()
+	if not x or not y:
+		return 0
+	sum = 0
+	for i in range(len(x)):
+		sum += (x[i] - y[i])**2
+	return sum**(1/2)
 
 def manhattan_dist(x, y):
-    raise NotImplementedError()
+	if not x or not y:
+		return 0
+	sum = 0
+	for i in range(len(x)):
+		if x[i] > y[i]:
+			sum += x[i] - y[i]
+		else:
+			sum += y[i] - x[i]
+	return sum
 
 def jaccard_dist(x, y):
-    raise NotImplementedError()
+	if not x or not y:
+		return 0
+	diff = 0
+	score = 0
+	for i in range(len(x)):
+		if x[i] != y[i]:
+			diff += 1
+		else:
+			score += 1
+	return 1 - score/(score+diff)
 
 def cosine_sim(x, y):
-    raise NotImplementedError()
+	if not x or not y:
+		return 0
+	vec_product = 0
+	for i in range(len(x)):
+		vec_product += x[i] * y[i]
+	norm_product = euclidean_dist(x, y)
+	if norm_product == 0:
+		return 1
```

Please raise an error instead here. Returning 1 implies that x and y are equal

---

solved!